db/async_wait.c

```diff
@@ -0,0 +1,869 @@
+                    item = item->lsn_lnk.next;
+                    Pthread_mutex_unlock(&(work_queue->mutex));
+                }
+            }
+        } else {
+            if (gbl_async_dist_commit_verbose) {
+                logmsg(LOGMSG_USER, "Timed out on conditional wait.. Going over absolute ts list\n");
+            }
+            Pthread_mutex_lock(&(work_queue->mutex));
+            item = LISTC_TOP(&(work_queue->absolute_ts_list));
+            Pthread_mutex_unlock(&(work_queue->mutex));
+            cur_time = comdb2_time_epochms();
+            while (item != NULL && (item->next_ts <= cur_time)) {
+                if (gbl_async_dist_commit_verbose) {
+                    logmsg(LOGMSG_USER, "Processing work item with seqnum  %d:%d, in state: %d\n",
+                           item->seqnum.lsn.file, item->seqnum.lsn.offset, item->cur_state);
+                }
+                process_work_item(item);
+                if (item->cur_state == DONE) {
+                    if (gbl_async_dist_commit_verbose) {
+                        logmsg(LOGMSG_USER, "Done processing work item with seqnum %d:%d... Freeing it\n",
+                               item->seqnum.lsn.file, item->seqnum.lsn.offset);
+                    }
+                    Pthread_mutex_lock(&(work_queue->mutex));
+                    next_item = item->absolute_ts_lnk.next;
+                    remove_from_async_wait_queue(item);
+                    Pthread_mutex_unlock(&(work_queue->mutex));
+                    deallocate_async_wait_node(item);
+                    item = next_item;
+                } else {
+                    Pthread_mutex_lock(&(work_queue->mutex));
+                    item = item->absolute_ts_lnk.next;
+                    Pthread_mutex_unlock(&(work_queue->mutex));
+                }
+            }
+            if (item != NULL) {
+                /* Before going into a timed wait check if we've got new lsns in the meanwhile.
+                   In this case, we shouldn't go into a timed wait... we should rather go over the lsn list */
+                if (local_new_lsns != new_lsns) {
+                    if (gbl_async_dist_commit_verbose) {
+                        logmsg(LOGMSG_USER, "Not going into timed wait as we got new lsns\n");
+                    }
+                    wait_rc = 0; /* since it's not ETIMEDOUT, we go through LSN list in next iteration */
+                    continue;
+                }
+                setup_waittime(&waittime, item->next_ts - comdb2_time_epochms());
+                if (gbl_async_dist_commit_verbose) {
+                    logmsg(LOGMSG_USER,
+                           "Current time(%d) before earliest time (%ld) a work item has to be checked. going into "
+                           "timedcondwait\n",
+                           comdb2_time_epochms(), item->next_ts);
+                }
+                Pthread_mutex_lock(&(item->bdb_state->seqnum_info->lock));
```

Comment:
check new_lsns == local_new_lsns again under this seqnum-mutex - only wait if they are equal